File: scratchpad/failed-jobs/2026-06-11__09-29-21/publish-api-bulk-100__JQvKp93/artifacts/user/hookdeck-task/verify.py
#!/usr/bin/env python3
"""Verify all 100 requests have correct bodies and headers, and all events are SUCCESSFUL."""

import json
import subprocess
import os
import sys
import time

API_BASE = "https://api.hookdeck.com/2025-07-01"
API_KEY = os.environ.get("HOOKDECK_API_KEY", "")
SOURCE_ID = "src_shf232fkafkqau"
RUN_ID = os.environ.get("ZEALT_RUN_ID", "")

def api_get(path, params=""):
    url = f"{API_BASE}{path}?{params}" if params else f"{API_BASE}{path}"
    result = subprocess.run(
        ["curl", "-s", url, "-H", f"Authorization: Bearer {API_KEY}"],
        capture_output=True, text=True
    )
    return json.loads(result.stdout)

def api_get_raw(path, params=""):
    url = f"{API_BASE}{path}?{params}" if params else f"{API_BASE}{path}"
    result = subprocess.run(
        ["curl", "-s", url, "-H", f"Authorization: Bearer {API_KEY}"],
        capture_output=True, text=True
    )
    return result.stdout

# 1. Verify requests count
print("=== Verifying Requests ===")
requests_data = api_get("/requests", f"source_id={SOURCE_ID}&limit=100")
request_models = requests_data.get("models", [])
print(f"Total requests found: {len(request_models)}")

# 2. Verify events are all SUCCESSFUL
print("\n=== Verifying Events ===")
events_data = api_get("/events", f"source_id={SOURCE_ID}&limit=100&status=SUCCESSFUL")
event_models = events_data.get("models", [])
print(f"Successful events found: {len(event_models)}")

# Also check for non-successful events
other_events = api_get("/events", f"source_id={SOURCE_ID}&limit=100&status=FAILED")
failed_count = len(other_events.get("models", []))
other_events2 = api_get("/events", f"source_id={SOURCE_ID}&limit=100&status=PENDING")
pending_count = len(other_events2.get("models", []))
print(f"Failed events: {failed_count}")
print(f"Pending events: {pending_count}")

# 3. Verify individual request bodies and headers
print("\n=== Verifying Request Bodies and Headers ===")
i_values = set()
batch_id_count = 0
errors = []

for req in request_models:
    req_id = req["id"]
    req_detail = json.loads(api_get_raw(f"/requests/{req_id}"))
    
    # Check body
    body = req_detail.get("data", {}).get("body", {})
    if isinstance(body, str):
        body = json.loads(body)
    
    if "i" not in body:
        errors.append(f"Request {req_id}: missing 'i' in body: {body}")
    else:
        i_val = body["i"]
        if i_val in i_values:
            errors.append(f"Request {req_id}: duplicate i={i_val}")
        i_values.add(i_val)
    
    # Check x-batch-id header
    headers = req_detail.get("data", {}).get("headers", {})
    batch_id = headers.get("x-batch-id", "")
    if batch_id == "BATCH-001":
        batch_id_count += 1
    else:
        errors.append(f"Request {req_id}: x-batch-id={batch_id} (expected BATCH-001)")

print(f"Unique i values: {len(i_values)}")
print(f"i range: {min(i_values) if i_values else 'N/A'} - {max(i_values) if i_values else 'N/A'}")
print(f"Requests with x-batch-id: BATCH-001: {batch_id_count}")
print(f"i values match 0-99: {i_values == set(range(100))}")

if errors:
    print(f"\n=== Errors ({len(errors)}) ===")
    for e in errors[:10]:
        print(f"  {e}")
else:
    print("\n✅ All verifications passed!")

# 4. Write summary
print("\n=== Summary ===")
print(f"Source Name: bulk-source-{RUN_ID}")
print(f"Destination Name: bulk-dest-{RUN_ID}")
print(f"Published Count: 100")
print(f"Batch ID: BATCH-001")

File: scratchpad/failed-jobs/2026-06-11__08-14-38/publish-api-bulk-100__VupydDK/artifacts/user/hookdeck-task/verify.js
const apiKey = process.env.HOOKDECK_API_KEY;
const runId = process.env.ZEALT_RUN_ID;

if (!apiKey || !runId) {
  console.error("Missing HOOKDECK_API_KEY or ZEALT_RUN_ID environment variables.");
  process.exit(1);
}

const sourceId = "src_2qt8dcr5rj6080"; // Handled dynamically if we want, but we can also fetch it by name or use the ID we know.
const sourceName = `bulk-source-${runId}`;

async function getSourceId() {
  const url = `https://api.hookdeck.com/2025-07-01/sources`;
  const res = await fetch(url, {
    headers: { "Authorization": `Bearer ${apiKey}` }
  });
  const data = await res.json();
  const source = data.models.find(s => s.name === sourceName);
  if (!source) {
    throw new Error(`Source not found: ${sourceName}`);
  }
  return source.id;
}

async function verify() {
  const sId = await getSourceId();
  console.log(`Using Source ID: ${sId} for verification.`);

  // Fetch all events for this source
  let url = `https://api.hookdeck.com/2025-07-01/events?source_id=${sId}&limit=100`;
  let allEvents = [];
  
  while (url) {
    const res = await fetch(url, {
      headers: { "Authorization": `Bearer ${apiKey}` }
    });
    const data = await res.json();
    allEvents.push(...data.models);
    
    if (data.pagination && data.pagination.next) {
      url = `https://api.hookdeck.com/2025-07-01/events?source_id=${sId}&limit=100&next=${data.pagination.next}`;
    } else {
      url = null;
    }
  }

  console.log(`Total events fetched: ${allEvents.length}`);

  // Fetch full details for each event to get body and headers
  const batchSize = 10;
  const verifiedEvents = [];
  
  for (let start = 0; start < allEvents.length; start += batchSize) {
    const batch = allEvents.slice(start, start + batchSize);
    const promises = batch.map(async (evt) => {
      const detailUrl = `https://api.hookdeck.com/2025-07-01/events/${evt.id}`;
      const res = await fetch(detailUrl, {
        headers: { "Authorization": `Bearer ${apiKey}` }
      });
      if (!res.ok) {
        throw new Error(`Failed to fetch event details for ${evt.id}`);
      }
      return await res.json();
    });
    
    const details = await Promise.all(promises);
    verifiedEvents.push(...details);
    console.log(`Fetched details for ${verifiedEvents.length}/${allEvents.length} events...`);
  }

  // Verify the requirements
  let successfulCount = 0;
  let batchIdCount = 0;
  const receivedValues = new Set();
  const duplicateValues = [];
  const extraValues = [];

  for (const evt of verifiedEvents) {
    if (evt.status === "SUCCESSFUL") {
      successfulCount++;
    }

    const headers = evt.data?.headers || {};
    // Header keys are case-insensitive, but Hookdeck usually normalizes them or returns them as they were sent.
    // Let's check both 'x-batch-id' and 'X-Batch-Id'.
    const batchId = headers["x-batch-id"] || headers["X-Batch-ID"] || headers["X-Batch-Id"];
    if (batchId === "BATCH-001") {
      batchIdCount++;
    }

    const body = evt.data?.body;
    if (body && typeof body.i === "number") {
      const val = body.i;
      if (val >= 0 && val < 100) {
        if (receivedValues.has(val)) {
          duplicateValues.push(val);
        } else {
          receivedValues.add(val);
        }
      } else {
        extraValues.push(val);
      }
    }
  }

  console.log("\n--- VERIFICATION RESULTS ---");
  console.log(`SUCCESSFUL Status Count: ${successfulCount}`);
  console.log(`x-batch-id: BATCH-001 Count: ${batchIdCount}`);
  console.log(`Unique values count (0-99): ${receivedValues.size}`);
  console.log(`Duplicate values:`, duplicateValues);
  console.log(`Extra values:`, extraValues);

  const missingValues = [];
  for (let i = 0; i < 100; i++) {
    if (!receivedValues.has(i)) {
      missingValues.push(i);
    }
  }
  console.log(`Missing values:`, missingValues);

  if (
    successfulCount === 100 &&
    batchIdCount === 100 &&
    receivedValues.size === 100 &&
    duplicateValues.length === 0 &&
    extraValues.length === 0 &&
    missingValues.length === 0
  ) {
    console.log("\n✅ ALL ACCEPTANCE CRITERIA MET SUCCESSFULLY!");
  } else {
    console.log("\n❌ SOME CRITERIA FAILED!");
    process.exit(1);
  }
}

verify().catch(err => {
  console.error("Verification failed:", err);
  process.exit(1);
});
